dev/set_matrix.py

```diff
@@ -0,0 +1,427 @@
+    ['foo']
+    >>> process_requirements({"== dev": ["foo"]}, "dev")
+    ['foo']
+    >>> process_requirements({"> 0.1, != dev": ["foo"]}, "dev")
+    []
+    """
+    if requirements is None:
+        return []
+
+    if isinstance(requirements, list):
+        return requirements
+
+    dev_numeric = "9999.9999.9999"
+
+    if version == "dev":
+        version = dev_numeric
+
+    if isinstance(requirements, dict):
+        for ver_spec, packages in requirements.items():
+            op_and_ver_pairs = map(get_operator_and_version, ver_spec.split(","))
+            match_all = all(
+                op(LooseVersion(version), LooseVersion(dev_numeric if ver == "dev" else ver))
+                for op, ver in op_and_ver_pairs
+            )
+            if match_all:
+                return packages
+        else:
+            return []
+
+    raise TypeError("Should not reach here")
+
+
+def remove_comments(s):
+    """
+    Examples
+    --------
+    >>> code = '''
+    ... # comment 1
+    ...  # comment 2
+    ... echo foo
+    ... '''
+    >>> remove_comments(code)
+    'echo foo'
+    """
+    return "\n".join(l for l in s.strip().split("\n") if not l.strip().startswith("#"))
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--diff-only",
+        action="store_true",
+        help="If specified, ignore unchanged flavors or configs.",
+    )
+    # make `--ref-config` and `--diff-files` configurable to make it easier to test this script.
+    parser.add_argument(
+        "--ref-config",
+        default="https://raw.githubusercontent.com/mlflow/mlflow/master/{}".format(CONFIG_PATH),
+        help=(
+            "URL or local file path of the reference config. "
+            "Valid only when `--diff-only` is specified."
+        ),
+    )
+    parser.add_argument(
+        "--changed-files",
+        type=lambda x: [] if x.strip() == "" else x.strip().split("\n"),
+        default="",
+        help=(
+            "A string that represents a list of changed files in a pull request. "
+            "Valid only when `--diff-only` is specified."
+        ),
+    )
+
+    return parser.parse_args()
+
+
+def main():
+    args = parse_args()
+
+    print(divider("Parameters"))
+    print(json.dumps(vars(args), indent=2))
+
+    print(divider("Log"))
+    config = read_yaml(CONFIG_PATH)
+    try:
+        config_ref = read_yaml(args.ref_config)
+    except Exception as e:
+        print("Failed to read '{}' due to: '{}'".format(args.ref_config, e))
+        config_ref = {}
+
+    # assuming that the top-level keys in `ml-package-versions.yml` have the format:
+    # <flavor name>(-<suffix>) (e.g. sklearn, tensorflow-1.x, keras-tf1.x)
+    flavors = set(x.split("-")[0] for x in config.keys())
+    changed_flavors = get_changed_flavors(args.changed_files, flavors)
+
+    job_names = []
+    includes = []
+    should_include_all_items = not args.diff_only
+
+    for flavor, cfgs in config.items():
```

Comment:
What happens if `config` from a PR branch is out-of-date relative to `config-ref`? Seems like we skip things that were added in `config_ref` in that case. To simplify the logic here, can we use the file that has the most recent commit (the config on `master` or the config on the PR branch) as the source of truth?

Reply:
Actually, shouldn't the GitHub actions workflow be executed on a branch that merges the PR with `master`? In that case, isn't `config` already up to date with all of the changes from the PR along with the most recent content in `master`? If so, I don't see why we need to do this. I know we talked about this before, but the logic is still a bit lost on me.

Reply:
Edit: I see. We provide `config_ref` so that we can test  *additive* changes to the config in the PR branch without retesting the entire config. Can we add an inline comment explaining this?